### PR TITLE
Fix: Ensure writeback handles NO_SHARD correctly by flattening tensors before copying

### LINKED ISSUE #151223: [FSDP] Cannot writeback when the parameter shape changes
### 🐛 Describe the bug

```python
import os
import torch
from torch import nn
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP

cmd = """
CUDA_VISIBLE_DEVICES=0 torchrun --nproc-per-node=1 mini_bug_reproduce.py # raise error
CUDA_VISIBLE_DEVICES=0,1 torchrun --nproc-per-node=2 mini_bug_reproduce.py # pass
"""

torch.distributed.init_process_group(backend='nccl')
world_size = int(os.environ["WORLD_SIZE"])
local_rank = int(os.environ["LOCAL_RANK"])
torch.set_default_device(local_rank)
device = torch.get_default_device()

class Test(nn.Module):

    def __init__(self):
        super().__init__()
        self.e = nn.Embedding(5, 4)

    def forward(self, x):
        x = self.e(x)
        return x.sum()

model = Test().half()

model = FSDP(
    model,
    use_orig_params=True,
)

## copied from https://github.com/huggingface/accelerate/blob/main/src/accelerate/accelerator.py#L1679-1719
upcasted_log = []
for module in FSDP.fsdp_modules(model):
    if not module._has_params:
        continue  # skip if FSDP module not managing parameters
    param = module._flat_param
    if (
        param.dtype != torch.float32
        and param.device != torch.device("meta")
        and param.requires_grad
    ):
        # keep log of names_params that was upcasted
        # NOTE: resorted to this because warnings.simplefilter("once") is somehow not working
        name_param_log = (module.module.__class__.__name__, ", ".join(module._flat_param._fqns))
        if name_param_log not in upcasted_log:
            upcasted_log.append(name_param_log)

        # this works because of FSDP's _runtime_utils.lazy_init.
        # Have to be careful not to call anything before this that
        # triggers lazy_init (e.g., _is_fsdp_root).
        param.data = param.data.to(torch.float32)  # upcasting
        module._handle._orig_param_dtype = torch.float32  # update

x = torch.randint(0, 5, (20,), device=device)

model.eval()
with torch.no_grad():
    loss = model(x)
```

1 gpu raise error, 2 gpus pass
```log
Expects torch.Size([20]) but got torch.Size([5, 4])
```

It seems that proper initialization do not occur when using only a single GPU.

### Versions

torch 2.6.0

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

### PR DESCRIPTION
**Pull request by ccchow**
Fixes #151223 

Because FSDP stores original parameters as views into a flattened tensor, changing the flattened parameter’s tensor directly can desynchronize the views. With the NO_SHARD strategy this caused a shape mismatch error when writing back modified parameters.

Ensured writeback handles NO_SHARD correctly by flattening tensors before copying. The logic now flattens the source parameter or gradient when the strategy is unsharded to maintain the expected 1‑D shape for writeback operations

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

### CODE REVIEW

**Review by weifengpy — APPROVED** [review]
thanks for the fix! are you interested in adding a unit test to cover this case? If not, I can follow up after your landing

**Comment by ccchow** [comment]
Sure I’ll add UT for this.

Best Regards,
*Lei*

**Comment by ccchow** [comment]
Hi @weifengpy , I added a case to test_fsdp_flatten_params.py. Please help review.

**Review by weifengpy — APPROVED** [review]

**Comment by ccchow** [comment]
Also confirmed that this fix resolves the repro in the original issue #151223

**Comment by ccchow** [comment]
Removed trailing spaces reported by lintrunner. @weifengpy could you help kick off merge again?

**Comment by ccchow** [comment]
Kindly ping.

**Comment by weifengpy** [comment]
> Kindly ping.

thanks for the reminder. triggering CI again

**Comment by ccchow** [comment]
Ran lintrunner -a torch/distributed/fsdp/_flat_param.py to resolve check errors.

**Comment by weifengpy** [comment]
> Ran lintrunner -a torch/distributed/fsdp/_flat_param.py to resolve check errors.

thanks. approved running cI again

**Comment by ccchow** [comment]
Not sure what's wrong with the checks but they probably need restart.

**Comment by ccchow** [comment]
Seems still blocked due to PR workflow being cancelled.

**Comment by ccchow** [comment]
I'm wondering how to resolve the pending workflows to unblock this PR.

**Comment by weifengpy** [comment]
> I'm wondering how to resolve the pending workflows to unblock this PR.

kicking it off again. 

if it fails, would it be possible to rebase on latest trunk? it’s 22 days old. if it succeeded, no need to rebase

**Comment by ccchow** [comment]
Thanks @weifengpy . Rebased my branch and should be ready to merge.

**Comment by weifengpy** [comment]
> Thanks @weifengpy . Rebased my branch and should be ready to merge.

running CI again

**Comment by ccchow** [comment]
Hi @weifengpy , the last CI got 100% passed but the PR is still blocked with error 'Merging is blocked
You're not authorized to push to this branch.'. I rebased again and appended a merge command. Feel free to trigger CI again and let me know if anything is needed.

**Comment by ccchow** [comment]
Thanks @weifengpy for the help on this PR!

**Comment by weifengpy** [comment]
appreicate your persistence to push to the very last mile @ccchow
